internal/test/gadgetapi.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/stretchr/testify/require"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	igjson "github.com/inspektor-gadget/inspektor-gadget/pkg/datasource/formatters/json"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/localmanager"
+	ocihandler "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/oci-handler"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/simple"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/socketenricher"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/local"
+)
+
+type GadgetApi[T any] struct {
+	image         string
+	timeout       time.Duration
+	mnsFilterMap  *ebpf.Map
+	apiParams     api.ParamValues
+	runtimeParams *params.Params
+
+	gadgetCtx      *gadgetcontext.GadgetContext
+	gadgetOperator operators.DataOperator
+	DataFunc       datasource.DataFunc
+	DataOperator   []operators.DataOperator
+	JsonFormatter  *igjson.Formatter
+
+	CapturedEvent   []T
+	OnGadgetRun     func(gadgetCtx operators.GadgetContext) error
+	BeforeGadgetRun func() error
+	NormalizeField  func(event *T)
+}
+
+// NewGadget initializes a new GadgetApi
+func NewGadget[T any](imageName, imageVersion string, timeout time.Duration, apiParams api.ParamValues, eventtype T, mnsFilterMap *ebpf.Map) (*GadgetApi[T], error) {
+	if imageName == "" {
+		return nil, fmt.Errorf("image name is required")
+	}
+	if imageVersion == "" {
+		imageVersion = "latest"
+	}
+	if timeout < 0 {
+		return nil, fmt.Errorf("invalid timeout value")
+	}
+
+	imageURL := fmt.Sprintf("ghcr.io/inspektor-gadget/gadget/%s:%s", imageName, imageVersion)
```

Comment:
do not append the specific inspektor gadget repo, this should be usable by external gadgets as well.